prometheus-ksonnet/lib/prometheus-configmap.libsonnet

```diff
@@ -1,28 +1,47 @@
+local util = import '_util.libsonnet';
+
 {
   prometheus+:: {
     name:: error 'must specify name',
 
     // We bounce through various layers of indirection so user can:
     // a) override config for all Prometheus' by merging into $.prometheus_config,
     // b) override config for a specific Prometheus instance by merging in here.
-    prometheus_config:: $.prometheus_config,
+    prometheus_config:: $.prometheus_config {
+      rule_files+: std.flattenArrays([
+        [
+          '%s/alerts.rules' % util.normalise(mixinName),
+          '%s/recording.rules' % util.normalise(mixinName),
+        ]
+        for mixinName in std.objectFields($.mixins)
+      ]),
+    },
+
     prometheusAlerts:: $.prometheusAlerts,
     prometheusRules:: $.prometheusRules,
 
-    local configMap = $.core.v1.configMap,
+    local prometheus_config = self.prometheus_config,
+    local prometheusAlerts = self.prometheusAlerts,
+    local prometheusRules = self.prometheusRules,
 
-    prometheus_config_map:
-      // Can't reference self.foo below as we're in a map context, so
-      // need to capture reference to the configs in scope here.
-      local prometheus_config = self.prometheus_config;
-      local prometheusAlerts = self.prometheusAlerts;
-      local prometheusRules = self.prometheusRules;
+    local configMap = $.core.v1.configMap,
 
-      configMap.new('%s-config' % self.name) +
-      configMap.withData({
-        'prometheus.yml': $.util.manifestYaml(prometheus_config),
-        'alerts.rules': $.util.manifestYaml(prometheusAlerts),
-        'recording.rules': $.util.manifestYaml(prometheusRules),
-      }),
+    prometheus_config_maps: [
+                              configMap.new('%s-config' % self.name) +
+                              configMap.withData({
+                                'prometheus.yml': $.util.manifestYaml(prometheus_config),
+                                'alerts.rules': $.util.manifestYaml(prometheusAlerts),
+                                'recording.rules': $.util.manifestYaml(prometheusRules),
+                              }),
+                            ] +
+                            [
+                              local mixin = $.mixins[mixinName] + util.emptyMixin;
+                              configMap.new('%s-%s-config' % [self.name, util.normalise(mixinName)]) +
```

Comment:
Instead of looping over the mixins twice (here and in the configVolumeMount) and normalizing the name, you could do it once in a list and and use that list as an input for both? I assume it would cost less computing power and thus run faster.

Reply:
Aha, you already have that, in the `rule_files` on line 11 here.